goldmane/pkg/aggregator/index.go

```diff
@@ -0,0 +1,181 @@
+		// The flow key is different from the DiachronicFlow key at this index. Insert a new DiachronicFlow.
+		logrus.WithFields(logrus.Fields{
+			"flow": d,
+			"i":    index,
+		}).Debug("Inserting new DiachronicFlow into index")
+		idx.diachronics = append(idx.diachronics[:index], append([]*types.DiachronicFlow{d}, idx.diachronics[index:]...)...)
+	}
+	// The DiachronicFlow already exists in the index, so do nothing.
+}
+
+func (idx *index[E]) Remove(d *types.DiachronicFlow) {
+	// Find the index of the DiachronicFlow to be removed.
+	index := idx.lookup(d)
+
+	if index == len(idx.diachronics) {
+		// The DiachronicFlow doesn't exist in the index (and would have sorted to the end of the index).
+		// We can't remove a flow that doesn't exist, so log a warning and return.
+		logrus.WithFields(logrus.Fields{"flow": d}).Warn("Unable to remove flow - not found in index")
+		return
+	}
+
+	if idx.diachronics[index].Key == d.Key {
+		// The DiachronicFlow at the returned index is the same as the DiachronicFlow to be removed.
+		// Remove it from the index.
+		logrus.WithFields(logrus.Fields{
+			"flow": d,
+		}).Debug("Removing flow from index")
+		idx.diachronics = append(idx.diachronics[:index], idx.diachronics[index+1:]...)
+		return
+	} else {
+		// The DiachronicFlow at the returned index is not the same as the DiachronicFlow to be removed.
+		// This means the DiachronicFlow to be removed doesn't exist in the index.
+		logrus.WithFields(logrus.Fields{"flow": d, "i": idx.diachronics[index]}).Warn("Unable to remove flow - not found in index")
+	}
+}
+
+// lookup returns the index within the list of DiachronicFlows where the given DiachronicFlow either already exists or should be inserted.
+// - If the DiachronicFlow already exists, the index of the existing DiachronicFlow is returned.
+// - If the DiachronicFlow does not exist, the index where it should be inserted is returned.
+func (idx *index[E]) lookup(d *types.DiachronicFlow) int {
+	return sort.Search(len(idx.diachronics), func(i int) bool {
+		// Compare the new DiachronicFlow with the DiachronicFlow at index i.
+		// - If the new DiachronicFlow sorts before the current DiachronicFlow, return true.
+		// - If the new DiachronicFlow sorts after the current DiachronicFlow, return false.
+		// - If this flow sorts the same as the current DiachronicFlow based on the parameters of this Index,
+		//   we need to sort based on the entire flow key to find a deterministic order.
+		// on the entire flow key.
+		k1 := idx.keyFunc(&idx.diachronics[i].Key)
+		k2 := idx.keyFunc(&d.Key)
+		if k1 > k2 {
+			// The key of the DiachronicFlow at index i greater than the key of the flow.
+			return true
+		}
+		if k1 == k2 {
+			// The field(s) this Index is optimized for considers these keys the same.
+			// Sort based on the key's ID to ensure a deterministic order.
+			// TODO: This will result in different ordering on restart. Should we sort by FlowKey fields instead
```

Comment:
It might not actually matter if the order changes after restart, since this is just for efficient lookups. The only thing that would be affected is the slight time difference certain lookups take for specific flows, i.e. looking up a flow that is in middle of the or ends of the index will be found in 1 or 2 checks, but that could move to somewhere that takes logn checks.